numpy/tests/test_public_api.py

```diff
@@ -739,3 +739,61 @@ def test___module__attribute():
 
     if incorrect_entries:
         assert len(incorrect_entries) == 0, incorrect_entries
+
+
+def _check___qualname__(obj, module_name) -> bool:
+    qualname = obj.__qualname__
+    name = obj.__name__
+
+    if "numpy.random" in module_name:
+        if type(getattr(obj, "__self__", None)) is np.random.RandomState:
+            return qualname == f"RandomState.{name}"
+        elif obj is np.random.bit_generator.randbits:
+            return qualname == f"SystemRandom.{name}"
+        else:
+            return qualname == name
+    elif "f2py" in module_name and name in ("match", "search"):
+        return qualname == f"Pattern.{name}"
+    elif "ma" in module_name and name == "reduce":
+        return qualname == f"_MaskedBinaryOperation.{name}"
+    else:
+        return qualname == name
```

Comment:
How come you don't use a pattern like
```
# Check that the/an object exists at `__module__:__qualname__`.
module = sys.modules[module_name]
actual_obj = functools.reduce(getattr, obj.__qualname__.split("."), module)

# Bound methods will not match object identity, so also accept qualname match
assert actual_obj is obj or actual_obj.__qualname__ == obj.__qualname__
```
or similar?  I think that does the same check effectively?  Could additionally check for `__name__ == qualname.split(".")[-1]`.

I suppose partly, this test (either way?) relies on already checking that there are no bad aliases flying around in `test_functions_single_location`.